### DISCUSSION TITLE
freezing nested packages (with relative imports)

### DISCUSSION THREAD

**Discussion by iot49** [micropython/micropython#11703] [General]
I'm trying to freeze a package with a sub-package, i.e.

```
package_top/
    __init__.py
   a.py
   package_nest/
       __init__.py
       b.py
```

My `manifest.py` looks as follows:

```
package("package_top")
```

The following do not work:

```
package_top/__init__.py:
     from . import package_nest
     package_nest.b. ...
```

My questions:

1. How can I import a nested package (`package_nest`)?
2. How can I use relative imports in frozen packages (`from . import ...`, etc)?

Of course it's possible to accomplish all this without nested packages or relative imports. Just the namespace get a bit more cluttered.

P.S. Without freezing everything works as expected. Also, other (not nested) packages and module work as expected after freezing.

**Comment by jimmo** [comment]
@iot49 Freezing should make no difference to the behavior of the code. Anything that works on the filesystem should work identically when it's frozen.

What you have should work (and works for me, both frozen and filesystem). I will note that `package_top/package_nest/__init__.py` must contain `from . import b` in order to be able to use `package_nest.b`. Or you must `import package_nest.b` in `package_top/__init__.py` (That's true regardless of filesystem or frozen).

**Reply by iot49** [reply]
Odd. When I tried 'from . import b', b was not found. Do I need a separate @package for 'b'?  If so, what name and path would I specify?

**Reply by jimmo** [reply]
@iot49 I think you'll need to share exactly the contents of your files that shows the problem.

**Reply by iot49** [reply]
The repo is at `https://github.com/iot49/rv-gateway`. 

`code-frozen` contains `manifest.py` and the code frozen into the custom micropython VM. 
`code` contains the "offending" package `app`. When I freeze it, micropython balks at 

```
from . import tasks
```

in file `app/__init__.py`. I.e. micropython finds all frozen code **except** the nested package `app/tasks`. 

Without freezing everything works as expected. Note that I have `package(app)` in the manifest, but no `package(app/tasks)` or whatever the correct syntax would be if such a directive is required. 

When I tried freezing `app`, that folder was also in the `code-frozen` folder. I just do not freeze it right now since it's not working. 

Further, I have changed `sys.path` in `boot.py` to first search the filesystem and `.frozen` last so I can override code for debugging and bug fixing. 

**In summary** everything works as expected _except_ freezing the nested frozen package `app/tasks`.

**Reply by jimmo** [reply]
Sorry @iot49 is there any way you can try to trim this down to a simpler example?

**Reply by iot49** [reply]
Will do, but will take some time since I do not have access to a computer right now (nor do I need the solution right now, obviously).
